pkg/frontend/querymiddleware/shard_active_native_histogram_metrics.go

```diff
@@ -0,0 +1,231 @@
+	defer spanLog.Finish()
+
+	resp, err := s.shardBySeriesSelector(ctx, spanLog, r, s.mergeResponses)
+
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func (s *shardActiveNativeHistogramMetricsMiddleware) mergeResponses(ctx context.Context, responses []*http.Response, encoding string) *http.Response {
+	reader, writer := io.Pipe()
+
+	mtx := sync.Mutex{}
+	metricIdx := make(map[string]int, 0)
+	metricBucketCount := make([]*cardinality.ActiveMetricWithBucketCount, 0)
+
+	updateMetric := func(item *cardinality.ActiveMetricWithBucketCount) {
+		if item == nil || len(item.Metric) == 0 {
+			// Skip empty/unknown metrics.
+			return
+		}
+		mtx.Lock()
+		defer mtx.Unlock()
+		if idx, ok := metricIdx[item.Metric]; ok {
+			metricBucketCount[idx].SeriesCount += item.SeriesCount
+			metricBucketCount[idx].BucketCount += item.BucketCount
+			if item.MinBucketCount < metricBucketCount[idx].MinBucketCount {
+				metricBucketCount[idx].MinBucketCount = item.MinBucketCount
+			}
+			if item.MaxBucketCount > metricBucketCount[idx].MaxBucketCount {
+				metricBucketCount[idx].MaxBucketCount = item.MaxBucketCount
+			}
+		} else {
+			metricIdx[item.Metric] = len(metricBucketCount)
+			metricBucketCount = append(metricBucketCount, item)
+		}
+	}
+
+	g := new(errgroup.Group)
+	for _, res := range responses {
+		if res == nil {
+			continue
+		}
+		r := res
+		g.Go(func() error {
+			defer func(body io.ReadCloser) {
+				// drain body reader
+				_, _ = io.Copy(io.Discard, body)
+				_ = body.Close()
+			}(r.Body)
+
+			bufPtr := jsoniterBufferPool.Get().(*[]byte)
+			defer jsoniterBufferPool.Put(bufPtr)
+
+			it := jsoniter.ConfigFastest.BorrowIterator(*bufPtr)
+			it.Reset(r.Body)
+			defer func() {
+				jsoniter.ConfigFastest.ReturnIterator(it)
+			}()
+
+			// Iterate over fields until we find data or error fields
+			foundDataField := false
+			for it.Error == nil {
+				field := it.ReadObject()
+				if field == "error" {
+					return fmt.Errorf("error in partial response: %s", it.ReadString())
+				}
+				if field == "data" {
+					foundDataField = true
+					break
+				}
+				// If the field is neither data nor error, we skip it.
+				it.ReadAny()
+			}
+			if !foundDataField {
+				return fmt.Errorf("expected data field at top level, found %s", it.CurrentBuffer())
+			}
+
+			if it.WhatIsNext() != jsoniter.ArrayValue {
+				err := errors.New("expected data field to contain an array")
+				return err
+			}
+
+			for it.ReadArray() {
+				if err := ctx.Err(); err != nil {
+					if cause := context.Cause(ctx); cause != nil {
+						return fmt.Errorf("aborted streaming because context was cancelled: %w", cause)
+					}
+					return ctx.Err()
+				}
+
+				item := cardinality.ActiveMetricWithBucketCount{}
+				it.ReadVal(&item)
+				updateMetric(&item)
+			}
+
+			return it.Error
+		})
+	}
+
+	// Cannot start streaming until we merged all results.
+	err := g.Wait()
```

Comment:
If we build the full response in memory then we don't need streaming at all, or am I missing something?

Reply:
I'm not sure about the efficiency. In general I think the non-streaming version is more robust because if there's any error during accumulation of the response the client will get a non-200 HTTP status code whereas with streaming the client will always get a 200 and if the response can't be built without error the stream will abort. Also, there would be less code complexity without the streaming. But since we've already got that complexity elsewhere maybe it's ok to duplicate it.

Reply:
Hmm, the error handling is more complicated in my case. That convinced me, I'll change to non streaming.

Reply:
Reworked it. I also switched from supporting  "x-snappy-framed" to supporting "snappy" as encoding, since framing is for streaming and we don't need it here.